### PR TITLE
[web_ui] Check if a pointer is already down for the specific device

### PR DESCRIPTION
**Pull request by hterkelsen**
Currently, we check if a pointer is down by checking if any device has sent a `down` event and we haven't received a corresponding `up` event. However, the check isn't taking into account the device id, so if you have one finger down, and press again with another finger, then it will synthesize an `up` event for the new finger, which crashes in the Flutter framework.

Change the `isButtonDown` check to keep a map of button-down status per-device.

### CODE REVIEW

File: lib/web_ui/lib/src/engine/pointer_binding.dart [hunk 1]
```diff
@@ -108,13 +108,15 @@ abstract class BaseAdapter {
 
   final DomRenderer domRenderer;
   PointerDataCallback _callback;
-  Map<int, bool> _isDownMap = <int, bool>{};
-  bool _isButtonDown(int button) {
-    return _isDownMap[button] == true;
+  Map<int, Map<int, bool>> _isDownMap = <int, Map<int,bool>>{};
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/pointer_binding.dart]: Let's add a comment that explains what we put in this map of maps

File: lib/web_ui/lib/src/engine/pointer_binding.dart [hunk 2]
```diff
@@ -108,13 +108,15 @@ abstract class BaseAdapter {
...
+  bool _isButtonDown(int device, int button) {
+    _isDownMap.putIfAbsent(device, () => <int, bool>{});
```
Comment (mdebbar) [inline lib/web_ui/lib/src/engine/pointer_binding.dart]: One concern I have here is the map will keep growing indefinitely. Especially in browsers that support `PointerEvent`s and provide a different `pointerId` for each new touch.

Comment (hterkelsen) [reply lib/web_ui/lib/src/engine/pointer_binding.dart]: How about keeping a set of (device_id, button) pairs: being in the set means that the device_id has that button down. When the button comes up then we take it out of the set

Comment (mdebbar) [reply lib/web_ui/lib/src/engine/pointer_binding.dart]: Yep, I think that would work!

File: lib/web_ui/lib/src/engine/pointer_binding.dart (hunk 1 above)
Comment (hterkelsen) [reply lib/web_ui/lib/src/engine/pointer_binding.dart]: Done.

File: lib/web_ui/lib/src/engine/pointer_binding.dart (hunk 2 above)
Comment (hterkelsen) [reply lib/web_ui/lib/src/engine/pointer_binding.dart]: Done.

**Comment by hterkelsen** [comment]
I added a test, PTAL

File: lib/web_ui/lib/src/engine/pointer_binding.dart [hunk 3]
```diff
@@ -129,7 +154,7 @@ abstract class BaseAdapter {
   void clearListeners() {
     final html.Element glassPane = domRenderer.glassPaneElement;
     _listeners.forEach((String eventName, html.EventListener listener) {
-      glassPane.removeEventListener(eventName, listener);
+      glassPane.removeEventListener(eventName, listener, true);
```
Comment (mdebbar) [inline lib/web_ui/lib/src/engine/pointer_binding.dart]: Nice catch!

**Pull request merged**